### PR TITLE
fastgltf: Add new recipe

### PR DESCRIPTION
**Pull request by spnda**
Specify library name and version:  **fastgltf/0.4.0**

I am the author of `fastgltf` and I want to add the recipe. This is a conan v2 script, and I am not sure how I should specify that the script won't work with 1.x? Also, should I still add the conanfile.py into my repository?

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

(Not 100% sure on that last point, but I think I did.)

### CODE REVIEW

**Comment by prince-chrismc** [comment]
> This is a conan v2 script, and I am not sure how I should specify that the script won't work with 1.x?

Conan 1. is still required at this time, there's a ton of enterprise users who are still migrating and offering recipes to them is advantageous for everyone :)

> Also, should I still add the conanfile.py into my repository?

No? There's not a strong need unless you plan on hosting your own packages -- also the source/exports would be different so it does not make much sense.

File: recipes/fastgltf/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            del self.options.fPIC
+
+    def layout(self):
+        cmake_layout(self, src_folder='src')
+
+    def validate(self):
+        check_min_cppstd(self, "17")
+
+    def requirements(self):
+        self.requires("simdjson/3.1.1")
```
Comment (prince-chrismc) [inline recipes/fastgltf/all/conanfile.py:44]: https://github.com/conan-io/conan-center-index/pull/16949#issuecomment-1501159000 new versions has pakcages for v2 :)

```suggestion
        self.requires("simdjson/3.1.7")
```

File: recipes/fastgltf/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def layout(self):
+        cmake_layout(self, src_folder='src')
+
+    def validate(self):
+        check_min_cppstd(self, "17")
```
Comment (prince-chrismc) [inline recipes/fastgltf/all/conanfile.py:41]: The 1.x pipeline does not user cppstd so it's not actually doing anything at the comment

You can see the FAQ which explains in more details :)

```suggestion
        if self.settings.get_safe("compiler.cppstd"):
            check_min_cppstd(self, "17")
```

File: recipes/fastgltf/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    # Needed for CMake configurations for dependencies to work
+    generators = "CMakeDeps"
+
+    settings = "os", "compiler", "build_type", "arch"
+
+    exports_sources = "CMakeLists.txt", "src/*", "cmake/*"
+
+    options = {
+        "fPIC": [True, False],
+        "useSmallVector": [True, False],
```
Comment (prince-chrismc) [inline recipes/fastgltf/all/conanfile.py:23]: As per the recommended naming (not required but it's easier for consumer when they are consistent -- requires less thinking and less coffee)

```suggestion
        "enable_small_vector": [True, False],
```

File: recipes/fastgltf/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    homepage = "https://github.com/spnda/fastgltf"
+    description = "A blazing fast C++17 glTF 2.0 library powered by SIMD."
+    topics = ("gltf", "simd", "cpp17")
+
+    # Needed for CMake configurations for dependencies to work
+    generators = "CMakeDeps"
+
+    settings = "os", "compiler", "build_type", "arch"
+
+    exports_sources = "CMakeLists.txt", "src/*", "cmake/*"
```
Comment (prince-chrismc) [inline recipes/fastgltf/all/conanfile.py:19]: None of these files are present 🤔 any reason for this line?

File: recipes/fastgltf/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    license = "MIT"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/spnda/fastgltf"
+    description = "A blazing fast C++17 glTF 2.0 library powered by SIMD."
+    topics = ("gltf", "simd", "cpp17")
+
+    # Needed for CMake configurations for dependencies to work
+    generators = "CMakeDeps"
+
+    settings = "os", "compiler", "build_type", "arch"
```
Comment (prince-chrismc) [inline recipes/fastgltf/all/conanfile.py:17]: For 2.0 to be even smarter

```suggestion
    package_type = "library"
    settings = "os", "compiler", "build_type", "arch"
```

**Review by prince-chrismc — COMMENTED** [review]
Overall this looks super good, just a few small questions and minor suggestions

File: recipes/fastgltf/all/conanfile.py (hunk 4 above)
Comment (prince-chrismc) [reply recipes/fastgltf/all/conanfile.py]: If the files are not in this repository then nothing matches these patterns

**Comment by prince-chrismc** [comment]
Please do not force push 🙏 GitHub forces us to restart the review which is not fun!

**Comment by prince-chrismc** [comment]
> ======== Computing dependency graph ========
> ERROR: Package type is 'library', but no 'shared' option declared

IS there a reason to skip it?

I see the code should allow for it to be managed by Conan https://github.com/spnda/fastgltf/blob/7758e62787ce7ea2e2b8e20ced7e0490fca28210/src/CMakeLists.txt#L1

**Comment by spnda** [comment]
> Please do not force push 🙏 GitHub forces us to restart the review which is not fun!

Is this specific to some scripted workflow? You can actually view the diff between force-pushes on GitHub, so you can see what I changed. But I will not force-push anymore.

> IS there a reason to skip it?
>
> I see the code should allow for it to be managed by Conan https://github.com/spnda/fastgltf/blob/7758e62787ce7ea2e2b8e20ced7e0490fca28210/src/CMakeLists.txt#L1

I really didn't understand how Conan wants to have the CMake file be setup for a shared library. Could you please link a precise example? I also don't think my code is setup properly to work on Windows as a shared library, so I decided I would just skip it but couldn't find a way to tell Conan this.

**Comment by prince-chrismc** [comment]
Screenshot_20230419_195520_Chrome

**Comment by memsharded** [comment]
Hi @spnda 

I did a small contribution with a couple of minor necessary fixes in https://github.com/conan-io/conan-center-index/pull/16704/commits/5f65af8b5041732665f13b1a0b15cf736e31785d. With those I managed to build the package in Windows, except I also had to build dependencies locally. So if it fails in CI it could be related to missing binaries of dependencies, not a recipe issue. 

Also, by adding the ``shared`` option, I managed to build it as a shared library, and it is also working locally, so I guess it is not a problem and the library also works fine as a shared library too.

In short: everything is looking good from your side, I'll try to keep an eye on this to have it merged.

**Comment by uilianries** [comment]
@spnda Thank you for your contribution, please consider my pull request https://github.com/spnda/conan-center-index/pull/1.

**Comment by spnda** [comment]
I've kinda lost track of what is to-do with this PR but it looks like its "good" now? If so, that's good but I would first like to update this PR to target 0.5.0 first as I recently released that.

**Review by uilianries — APPROVED** [review]
LGTM

File: recipes/fastgltf/all/test_v1_package/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+from conans import ConanFile, CMake
```
Comment (uilianries) [inline recipes/fastgltf/all/test_v1_package/conanfile.py:1]: the `test_v1_package` is no longer required for new recipes: https://github.com/conan-io/conan-center-index/discussions/17406

But you can keep as it does not harm and does not affect the CI.

Comment (spnda) [reply recipes/fastgltf/all/test_v1_package/conanfile.py:1]: I didn't add it myself, so I might as well just remove it.

**Review by SSE4 — APPROVED** [review]
